src/app/login-button/login-button.component.spec.ts

```diff
@@ -0,0 +1,74 @@
+import {HarnessLoader} from '@angular/cdk/testing';
+import {TestbedHarnessEnvironment} from '@angular/cdk/testing/testbed';
+import {ComponentFixture, TestBed} from '@angular/core/testing';
+import {MatButtonModule} from '@angular/material/button';
+import {MatButtonHarness} from '@angular/material/button/testing';
+
+import {AuthService} from '../auth.service';
+import {AuthServiceStub} from '../testing/auth-service-stub';
+
+import {LoginButtonComponent} from './login-button.component';
+
+describe('LoginButtonComponent', () => {
+  let authServiceStub: AuthServiceStub;
+  let compiled: HTMLElement;
+  let component: LoginButtonComponent;
+  let fixture: ComponentFixture<LoginButtonComponent>;
+  let loader: HarnessLoader;
+
+  beforeEach(async () => {
+    authServiceStub = new AuthServiceStub();
+
+    await TestBed
+        .configureTestingModule({
+          imports: [
+            MatButtonModule,
+          ],
+          declarations: [LoginButtonComponent],
+          providers: [{provide: AuthService, useValue: authServiceStub}],
+        })
+        .compileComponents();
+  });
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(LoginButtonComponent);
+    component = fixture.componentInstance;
+    compiled = fixture.nativeElement;
+    loader = TestbedHarnessEnvironment.loader(fixture);
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should show a log out button and display the user\'s email', async () => {
+    const logInButton = await loader.getHarness(
+        MatButtonHarness.with({selector: '.login-button'}));
+    expect(await logInButton.getText()).toContain('Log Out');
+
+    const emailLabel = compiled.querySelector('.user-email-label');
+    if (!emailLabel) throw new Error('Failed to find user email label');
+    const expectedEmail = authServiceStub.getUserEmail();
+    expect(emailLabel.textContent).toContain(expectedEmail);
+  });
+
+  it('should show a login button if the user is not logged in', async () => {
+    authServiceStub.user = null;
+    fixture.detectChanges();
+
+    const logInButton = await loader.getHarness(
+        MatButtonHarness.with({selector: '.login-button'}));
+    expect(await logInButton.getText()).toContain('Log In');
+  });
+
+  it('should attempt to log in or out when the button is clicked', async () => {
+    spyOn(authServiceStub, 'logInOrOut');
+
+    const logInButton = await loader.getHarness(
+        MatButtonHarness.with({selector: '.login-button'}));
+    await logInButton.click();
+
+    expect(authServiceStub.logInOrOut).toHaveBeenCalled();
```

Comment:
It's occurring to me that this test should maybe show that logIn or logOut is called in the appropriate situation.  I could imagine doing this in two ways - separate the login and logout functionalities into two different buttons, or have the logInOrOut functionality in the authServiceStub do the real call that authService does, and mock out logIn and logOut to see which one is called by logInOrOut.

Reply:
I changed the test and split it into two to spy on the logIn/logOut functions.

I also changed the auth service slightly so the logIn function returns Promise<void> since I'm not using the result of the call anywhere.